### DISCUSSION TITLE
It seems that add and remove cannot activate ObservableCollection to notify changes. How to solve this problem?

### DISCUSSION THREAD

**Discussion by IronEaterBeast** [AvaloniaUI/Avalonia#16334] [Q&A] [answered]
My view model:
```
public partial class MainWindowViewModel : ViewModelBase
{
    public ObservableCollection<OneLineClass> _allContentOC { get; set; } = [];
    public ObservableCollection<OneLineClass> AllContentOC
    {
        get
        {
            return _allContentOC;
        }
        set
        {
            _allContentOC = value;
            this.RaisePropertyChanged(nameOf(AllContentOC));
        }
    }
}
```

Can activate:
```
var test = new MainWindowViewModel();
test.AllContentOC = [];
```

Cannot activate
`
test.AllContentOC.Add(new OneLineClass("test"));
`

Did I miss something?

**Comment by maxkatz6** [comment]
Which changes do you want to activate? Adding a new item to the array won't notify parent view model about any changes. Instead you need to subscribe on `AllContentOC.CollectionChanged` event.

**Reply by IronEaterBeast** [reply]
In my imagination, adding a new item should be the same as assigning and should enter the set method, then use RaisePropertyChanged to update the window display, but it doesn't seem to work that way, so I don't know how to handle this situation.

**Reply by IronEaterBeast** [reply]
if OneLineClass is another ViewModelBase like `public class OneLineClass : ViewModelBase` then AllContentOC.CollectionChanged will not work too.

**Reply by maxkatz6** [reply, marked as answer]
> In my imagination, adding a new item should be the same as assigning and should enter the set method

It's not how it works. Normally "Property changed" event only is raised when any actual property is changed.
But since you implement your own ViewModelBase, that's up to you. Still - you wrote no code to call RaisePropertyChanged when collection items are changed. Setter won't be executed in this situation obviously.

 > if OneLineClass is another ViewModelBase like public class OneLineClass : ViewModelBase then AllContentOC.CollectionChanged will not work too.

AllContentOC.CollectionChanged still will work, when collection is changed.
But if you want to subscribe on **property changes** of the OneLineClass, you need to subscribe on each item. 

PropertyChanged and CollectionChanged are standard .NET events, and that's how they work. Avalonia just follows the same rules.

PropertyChanged is about property of **this** class, where event is defined.
And CollectionChanged is about collection changes of **this** collection/class, where event is defined.
